src/System.Private.CoreLib/shared/System/IO/BinaryReader.cs

```diff
@@ -649,6 +597,46 @@ public virtual byte[] ReadBytes(int count)
             return result;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ReadOnlySpan<byte> InternalRead(int numBytes)
+        {
+            Debug.Assert(numBytes >= 2 && numBytes <= 16, "value of 1 should use ReadByte. value > 16 requires to change the minimal _buffer size");
+
+            if (_isMemoryStream)
+            {
+                // no need to check if _stream == null as we will never have null _stream when _isMemoryStream = true
+
+                // read directly from MemoryStream buffer
+                MemoryStream mStream = _stream as MemoryStream;
+                Debug.Assert(mStream != null, "_stream as MemoryStream != null");
+
+                return mStream.InternalReadSpan(numBytes);
```

Comment:
Seems like this would be a little better as:
```C#
Debug.Assert(_stream is MemoryStream);
return ((MemoryStream)_stream).InternalReadSpan(numBytes);
```
That way, if somehow we ended up with _stream being a non-null stream that was something else, we'd at least get a cast exception with details rather than a null reference exception, and if it does end up being null, we'll get the same null ref exception we would today.

Reply:
good idea. fixed.